File: algorithms/balance_a_binary_search_tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;

fn inorder(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
    let mut arr = Vec::new();
    if root.is_none() {
        return arr;
    }
    let mut st = VecDeque::new();
    let mut done = false;
    let mut iter = root;
    while !done {
        if iter.is_some() {
            let node = iter.clone();
            iter = iter.unwrap().borrow().left.clone();
            st.push_back(node);
        } else {
            if st.len() == 0 {
                done = true;
            } else {
                iter = st.pop_back().unwrap();
                let node = iter.clone();
                iter = iter.unwrap().borrow().right.clone();
                arr.push(node.unwrap().borrow().val);
            }
        }
    }
    arr
}

fn build(arr: &[i32]) -> Option<Rc<RefCell<TreeNode>>> {
    if arr.len() == 0 {
        return None;
    }
    if arr.len() == 1 {
        return Some(Rc::new(RefCell::new(TreeNode::new(arr[0]))));
    }
    let mid = arr.len() / 2;
    let mut node = Rc::new(RefCell::new(TreeNode::new(arr[mid])));
    node.borrow_mut().left = build(&arr[..mid]);
    node.borrow_mut().right = build(&arr[mid + 1..]);
    Some(node)
}

impl Solution {
    pub fn balance_bst(root: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        let arr = inorder(root);
        build(&arr)
    }
}

File: algorithms/medium/balance_a_binary_search_tree.kt
/**
 * Example: var ti = TreeNode(5) var v = ti.`val` Definition for a binary tree node. class
 * TreeNode(var `val`: Int) { var left: TreeNode? = null var right: TreeNode? = null }
 */
class Solution {
  fun inorder(root: TreeNode?, lst: ArrayList<Int>) {
    if (root == null) {
      return
    }
    inorder(root.left, lst)
    lst.add(root.`val`)
    inorder(root.right, lst)
  }

  fun build(l: Int, r: Int, lst: ArrayList<Int>): TreeNode? {
    if (l > r) {
      return null
    }

    val m = (l + r) / 2
    val node = TreeNode(lst[m])
    node.left = build(l, m - 1, lst)
    node.right = build(m + 1, r, lst)
    return node
  }

  fun balanceBST(root: TreeNode?): TreeNode? {
    val lst = ArrayList<Int>()
    inorder(root, lst)
    return build(0, lst.size - 1, lst)
  }
}
